### PR TITLE
[UIManager] Reverse order of _task_queue

### PR DESCRIPTION
**Pull request by zwim**
In real life most scheduled tasks are in the near future. Tasks being long in the future are seldom rescheduled. 

In the current implementation if a near future task is scheduled it is inserted at the beginning of the task queue, and therefore most of the queue has to be shifted backwards. When such a task gets consumed, the whole queue is shifted forwards again. So there is a lot of shifting. The only benefit of this implementation is, that `_checkTasks` only needs to check `_task_queue[1]`.

This PR reverses the sorting order of the task_queue, so that tasks scheduled in the near future are at the end of the queue -> less shifting on insertion and consumption.

Additionally, there is a small benchmark `taskqueue_bench.lua` which tries to mimicry some real life usage with scheduling and consume tasks. This benchmark can be run on the current master branch or this PR.

For benching on master I get:
```
Benchmark #1: ./kodev test front taskqueue_bench.lua
  Time (mean ± σ):      5.398 s ±  0.073 s    [User: 5.138 s, System: 0.218 s]
  Range (min … max):    5.323 s …  5.540 s    10 runs
```
and for running it on this PR:
```
Benchmark #1: ./kodev test front taskqueue_bench.lua
  Time (mean ± σ):      4.544 s ±  0.029 s    [User: 4.276 s, System: 0.226 s]
  Range (min … max):    4.489 s …  4.585 s    10 runs
```

Which is an improvement of 16% (not bad).

- - -
This change is [Reviewable](https://reviewable.io/reviews/koreader/koreader/9706)

### CODE REVIEW

File: spec/unit/taskqueue_bench.lua [hunk 1]
```diff
@@ -0,0 +1,78 @@
...
+        UIManager:schedule(now + i, noop)
+        UIManager:schedule(now + NB_TESTS - i, noop)
+    end
+
+    for i=1, NB_TESTS do
+        wait_until, now = UIManager:_checkTasks() -- luacheck: no unused
+    end
+end)
+
+describe("UIManager schedule more sophiticated benchmark", function()
```
Comment (Frenzie) [inline spec/unit/taskqueue_bench.lua:27]: *cough*

File: frontend/ui/uimanager.lua [hunk 2]
```diff
@@ -905,21 +911,24 @@ function UIManager:_checkTasks()
     -- Tasks due for execution might themselves schedule more tasks (that might also be immediately due for execution ;)).
     -- Flipping this switch ensures we'll consume all such tasks *before* yielding to input polling.
     self._task_queue_dirty = false
-    while self._task_queue[1] do
-        local task_time = self._task_queue[1].time
+    local task = self._task_queue[self._task_queue.len]
+    while task do
```
Comment (NiLuJe) [inline frontend/ui/uimanager.lua:915]: `while self._task_queue.len > 0`?

(Mainly to avoid checking `self._task_queue[0]`, because that would break the "this is an array" contract ;).

File: frontend/ui/uimanager.lua [hunk 3]
```diff
@@ -237,10 +237,11 @@ function UIManager:close(widget, refreshtype, refreshregion, refreshdither)
     end
 end
 
--- Schedule an execution task; task queue is in ascending order
+-- Schedule an execution task; task queue is in descending order
 function UIManager:schedule(sched_time, action, ...)
     local lo, hi = 1, #self._task_queue
-    -- Rightmost binary insertion
+    self._task_queue.len = hi + 1
```
Comment (NiLuJe) [inline frontend/ui/uimanager.lua:243]: I'd go w/ `n` to follow Lua traditions ;).

(And set it to 0 at declaration time).

Comment (zwim) [reply frontend/ui/uimanager.lua:243]: You mean something like `self._task_queue = {n = 0}`?

Comment (NiLuJe) [reply frontend/ui/uimanager.lua:243]: Yup!

Comment (NiLuJe) [reply frontend/ui/uimanager.lua:243]: Or, well, not  `self.`, as I mean the actual declaration in @ ~L39

**Comment by NiLuJe** [comment]
Makes perfect sense, I'll give it another look tomorrow as my brain's shot right now ;).

**Review by NiLuJe — CHANGES_REQUESTED** [review]
> Reviewed 5 of 5 files at r4, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/koreader/koreader/9706)* status: all files reviewed, 4 unresolved discussions (waiting on @zwim)

___
*[`frontend/ui/uimanager.lua` line 257 at r4](https://reviewable.io/reviews/koreader/koreader/9706#-NFdhk0R1ek8Wqi6gFmz:-NFdhk0R1ek8Wqi6gFn-:b-td0fru) ([raw file](https://github.com/koreader/koreader/blob/f1acfbb8d291dcaf2ccbb43870c96499fdfe2d3a/frontend/ui/uimanager.lua#L257)):*
> ```Lua
>         else
>             lo = mid + 1
>         end
> ```

It's been a week, and I'm confused again ;o).

I'd much rather have this written as

```
        if mid_time < sched_time then
            hi = mid - 1
        else
            lo = mid + 1
        end
```

for consistency with the stuff in `util.bsearch`, because it breaks my brain for 10 minutes every time ;p.



_Code quote:_
```Lua
        if sched_time >= mid_time then


            hi = mid - 1
        else
            lo = mid + 1
        end
```
___
*[`frontend/ui/uimanager.lua` line 349 at r4](https://reviewable.io/reviews/koreader/koreader/9706#-NFdjHy2Cpbw4vkPyXtL:-NFdjHy2Cpbw4vkPyXtM:b-9cohob) ([raw file](https://github.com/koreader/koreader/blob/f1acfbb8d291dcaf2ccbb43870c96499fdfe2d3a/frontend/ui/uimanager.lua#L349)):*
> ```Lua
> function UIManager:unschedule(action)
>     local removed = false
>     local task_queue_len = self._task_queue.n
> ```

Why go through a local copy here? Can't we work on `self._task_queue.n` directly?

IIRC, the starting point of a for loop is only evaluated once.
___
*[`frontend/ui/uimanager.lua` line 915 at r4](https://reviewable.io/reviews/koreader/koreader/9706#-NFdjp-h1n5NoT1GgmtP:-NFdjp-h1n5NoT1GgmtQ:b-kkv6mf) ([raw file](https://github.com/koreader/koreader/blob/f1acfbb8d291dcaf2ccbb43870c96499fdfe2d3a/frontend/ui/uimanager.lua#L915)):*
> ```Lua
>     self._task_queue_dirty = false
>     while self._task_queue.n > 0 do
>         local task = self._task_queue[self._task_queue.n]
> ```

I much preferred the tighter scoping of just keeping the time here, and getting `task` from `table.remove` in the inner branch ;).

**Review by zwim — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/koreader/koreader/9706)* status: all files reviewed, 4 unresolved discussions (waiting on @NiLuJe)

___
*[`frontend/ui/uimanager.lua` line 257 at r4](https://reviewable.io/reviews/koreader/koreader/9706#-NFdhk0R1ek8Wqi6gFmz:-NFdvxzA2QnW61olxXGj:b5a7si5) ([raw file](https://github.com/koreader/koreader/blob/f1acfbb8d291dcaf2ccbb43870c96499fdfe2d3a/frontend/ui/uimanager.lua#L257)):*
<details><summary><i>Previously, NiLuJe (NiLuJe) wrote…</i></summary><blockquote>

It's been a week, and I'm confused again ;o).

I'd much rather have this written as

```
        if mid_time < sched_time then
            hi = mid - 1
        else
            lo = mid + 1
        end
```

for consistency with the stuff in `util.bsearch`, because it breaks my brain for 10 minutes every time ;p.
</blockquote></details>


This does not work 
```
[  FAILED  ] ./spec/front/unit/uimanager_spec.lua @ 127: UIManager spec should insert new tasks with same times after existing tasks
```

I would not touch that part of the code, swapping the if end else branch does exactly what I want: swap the sorting.
___
*[`frontend/ui/uimanager.lua` line 349 at r4](https://reviewable.io/reviews/koreader/koreader/9706#-NFdjHy2Cpbw4vkPyXtL:-NFdwTdV9DESXx9M0R2M:buv3h2y) ([raw file](https://github.com/koreader/koreader/blob/f1acfbb8d291dcaf2ccbb43870c96499fdfe2d3a/frontend/ui/uimanager.lua#L349)):*
<details><summary><i>Previously, NiLuJe (NiLuJe) wrote…</i></summary><blockquote>

Why go through a local copy here? Can't we work on `self._task_queue.n` directly?

IIRC, the starting point of a for loop is only evaluated once.
</blockquote></details>

That's clear. I just thought the `task_queue_len = task_queue_en -1` would save something (but that's not measurable, esp. if only one task is removed. -> so I'll remove the local.
___
*[`frontend/ui/uimanager.lua` line 915 at r4](https://reviewable.io/reviews/koreader/koreader/9706#-NFdjp-h1n5NoT1GgmtP:-NFdwrS-3p0XD0RjPjLh:b-896fix) ([raw file](https://github.com/koreader/koreader/blob/f1acfbb8d291dcaf2ccbb43870c96499fdfe2d3a/frontend/ui/uimanager.lua#L915)):*
<details><summary><i>Previously, NiLuJe (NiLuJe) wrote…</i></summary><blockquote>

I much preferred the tighter scoping of just keeping the time here, and getting `task` from `table.remove` in the inner branch ;).
</blockquote></details>

Done.

**Review by NiLuJe — CHANGES_REQUESTED** [review]
> Reviewed 1 of 1 files at r5, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/koreader/koreader/9706)* status: all files reviewed, 2 unresolved discussions (waiting on @zwim)

___
*[`frontend/ui/uimanager.lua` line 257 at r4](https://reviewable.io/reviews/koreader/koreader/9706#-NFdhk0R1ek8Wqi6gFmz:-NFeA_DFEhNwFmut1czr:bqbrbor) ([raw file](https://github.com/koreader/koreader/blob/f1acfbb8d291dcaf2ccbb43870c96499fdfe2d3a/frontend/ui/uimanager.lua#L257)):*
<details><summary><i>Previously, zwim wrote…</i></summary><blockquote>


This does not work 
```
[  FAILED  ] ./spec/front/unit/uimanager_spec.lua @ 127: UIManager spec should insert new tasks with same times after existing tasks
```

I would not touch that part of the code, swapping the if end else branch does exactly what I want: swap the sorting.
</blockquote></details>

That's intriguing, because the two are logically identical.

FWIW, that test *intermittently* fails on my desktop, regardless of this change, so there may be something rotten in the state of Denmark regardless ;).

**Comment by NiLuJe** [comment]
Actually, not that one, this one:

```
./spec/front/unit/uimanager_spec.lua:147: Expected objects to be the same.
Passed in:
(string) 'foo'
Expected:
(string) 'bar'

stack traceback:
        ./spec/front/unit/uimanager_spec.lua:147: in function <./spec/front/unit/uimanager_spec.lua:127>
```

**Comment by NiLuJe** [comment]
> FWIW, that test _intermittently_ fails on my desktop, regardless of this change, so there may be something rotten in the state of Denmark regardless ;).

Oh, the shenanigans with a switch to 11s that were reverted is what actually fixed the test on my local machine.

TL;DR: That test is horribly brittle on (I would guess) fast host and/or coarse clocks, so it needs to be rethought ;).

**Review by NiLuJe — CHANGES_REQUESTED** [review]
> *[Reviewable](https://reviewable.io/reviews/koreader/koreader/9706)* status: all files reviewed, 2 unresolved discussions (waiting on @zwim)

___
*[`frontend/ui/uimanager.lua` line 257 at r4](https://reviewable.io/reviews/koreader/koreader/9706#-NFdhk0R1ek8Wqi6gFmz:-NFeHHGD5kahJc-kPz86:b7id7e7) ([raw file](https://github.com/koreader/koreader/blob/f1acfbb8d291dcaf2ccbb43870c96499fdfe2d3a/frontend/ui/uimanager.lua#L257)):*
<details><summary><i>Previously, NiLuJe (NiLuJe) wrote…</i></summary><blockquote>

That's intriguing, because the two are logically identical.

FWIW, that test *intermittently* fails on my desktop, regardless of this change, so there may be something rotten in the state of Denmark regardless ;).
</blockquote></details>

Actually, I'm an idiot, this should obviously have read `mid_time <= sched_time` ;o).

**Comment by NiLuJe** [comment]
This passes on my end:

```
        -- insert task "5s" between "now" and "10s"
        UIManager:schedule(now, "now");
        assert.are.same("now", UIManager._task_queue[1].action)
        UIManager:schedule(now + time.s(10), "10s");
        assert.are.same("10s", UIManager._task_queue[1].action)
        UIManager:schedule(now + time.s(5), "5s");
        assert.are.same("5s", UIManager._task_queue[2].action)

        -- insert task in place of "10s", as it'll expire later
        -- NOTE: Can't use 10, as time.now, which is used internally, may or may not have moved,
        --       depending on host's performance and clock granularity (especially if host is fast and/or COARSE is available).
        UIManager:scheduleIn(11, 'foo') -- is a bit later than "10s", as time.now() is used internally
        assert.are.same('foo', UIManager._task_queue[1].action)

        -- insert task in place of "10s", which was just shifted by foo
        UIManager:schedule(now + time.s(10), 'bar')
        assert.are.same('bar', UIManager._task_queue[2].action)

        -- insert task in place of "bar"
        UIManager:schedule(now + time.s(10), 'baz')
        assert.are.same('baz', UIManager._task_queue[2].action)

        -- insert task in plkace of "5s"
        UIManager:schedule(now + time.s(5), 'nix')
        assert.are.same('nix', UIManager._task_queue[5].action)
        -- "barba" replaces "nix"
        UIManager:scheduleIn(5, 'barba') -- is a bit later than "5s", as time.now() is used internally
        assert.are.same('barba', UIManager._task_queue[5].action)

        -- "mama is sheduled now and as such inserted in "now"'s place
        UIManager:schedule(now, 'mama')
        assert.are.same('mama', UIManager._task_queue[8].action)

        -- "papa" is shortly after "now", so inserted in its place
        -- NOTE: For the same reason as above, test this last, as time.now may not have moved...
        UIManager:nextTick('papa') -- is a bit later than "now"
        assert.are.same('papa', UIManager._task_queue[8].action)

        -- "letta" is shortly after "papa", so inserted in its place
        UIManager:tickAfterNext('letta')
        assert.are.same("function", type(UIManager._task_queue[8].action))
```

(Also, updated the comments, which were now completely wrong)

**Comment by NiLuJe** [comment]
I'm halfway wondering if it wouldn't make more sense to use something like

```
local idx
        local idx_by_action = function(action)
            for i, t in ipairs(UIManager._task_queue) do
                if t.action == action then
                    return i
                end
            end

            return nil
        end
```

instead of hard-coded constants...

(But, then again, this would have hidden away this particular issue).

**Comment by zwim** [comment]
> > FWIW, that test _intermittently_ fails on my desktop, regardless of this change, so there may be something rotten in the state of Denmark regardless ;).
> 
> Oh, the shenanigans with a switch to 11s that were reverted is what actually fixed the test on my local machine.

I don't understand that behavior. The later inserted task should be inserted in the queue so, that it gets executed later anyway (In this PR, the later task should have a lower index). 

> 
> TL;DR: That test is horribly brittle on (I would guess) fast hosts and/or coarse clocks, so it needs to be rethought ;).

But on the other side you got me thinking:
When we insert a task1 with `scheduleIn(0,...)` then another task2 with `nextTick` and at last a task3 with `scheduleIn(0,..)`. 
Execution order should be task1, task3, task2.

If our clock is too slow or the machine too fast we get the wrong execution order of task1, task2, task3 (as all have the same time).

So maybe we should change nextTick to make sure it is really executed on the _next_ clock tick
```
function UIManager:nextTick(action, ...)
    local next_tick_delay = 0
    if time.now() > self._now then
        -- If we are on a fast machine or if we have a slow clock, postpone to the next system clock tick
        next_tick_delay = time.getResolution()
    end

    return self:scheduleIn(next_tick_delay, action, ...)
end
```

**Comment by NiLuJe** [comment]
> So maybe we should change nextTick to make sure it is really executed on the _next_ clock tick

Nope, the function is very badly named. It's meant to be the next UI frame, it bears no relation to clocks at all ;).

**Comment by NiLuJe** [comment]
I can't remember if I might not be the one to have named it, in which case it's probably because I was thinking of HLDS ticks at the time.

Regardless, yeah, nextFrame would have been much clearer ;p.

**Comment by Frenzie** [comment]
I thought you were purposefully taking potshots at past @NiLuJe. ;-)

**Comment by zwim** [comment]
Ok, lets drop the nextTick modification :) 

But for the 11s vs 10s test case should really be solved with a 1ms sleep before 'scheduleIn(10', as this us what we want to test, dont we?

**Comment by NiLuJe** [comment]
> But the 11s vs 10s test case should really be solved with a 1ms sleep before 'scheduleIn(10', as this Is what we want to test, dont we?

I would assume yeah, as something as simple as adding a print somewhere in schedule was enough to make it pass more reliably ;).

**Comment by zwim** [comment]
@NiLuJe can you please test if https://github.com/koreader/koreader/pull/9706/commits/6f67c3d4f15667afe840f4fe2b8e88f322325436 solves your intermittently failing problem? If not, please increase the usleep value.

**Comment by NiLuJe** [comment]
Will try to do so later tonight ;).

**Comment by NiLuJe** [comment]
> @NiLuJe can you please test if [6f67c3d](https://github.com/koreader/koreader/commit/6f67c3d4f15667afe840f4fe2b8e88f322325436) solves your intermittently failing problem? If not, please increase the usleep value.

1µs doesn't work, but 1ms does. Which possibly tracks, as that's the resolution of my system's COARSE MONOTONIC clock, IIRC.

**Comment by NiLuJe** [comment]
> 1µs doesn't work, but 1ms does. Which possibly tracks, as that's the resolution of my system's COARSE MONOTONIC clock, IIRC.

Yup!

`COARSE MONO res:        1000000LL`

FWIW, that is exactly the value we use as the threshold to use COARSE or not, so using that in the tests *should* be safe ;).

**Review by NiLuJe — APPROVED** [review]
> Reviewed 1 of 2 files at r6, 1 of 1 files at r8, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/koreader/koreader/9706)* status: all files reviewed, 2 unresolved discussions (waiting on @zwim)

___
*[`spec/unit/uimanager_spec.lua` line 141 at r8](https://reviewable.io/reviews/koreader/koreader/9706#-NFoGBIO7kMSS2G-yeux:-NFoGBIO7kMSS2G-yeuy:b-7nvldv) ([raw file](https://github.com/koreader/koreader/blob/db4557741382fcf6192570a9b3ac61e0b77930c2/spec/unit/uimanager_spec.lua#L141)):*
> ```Lua
>         assert.are.same("5s", UIManager._task_queue[2].action)
> 
>         -- insert task in place of "10s", as it'll expire shortyl afer "10s"
> ```

nit: `shortly` ;).

**Review by NiLuJe — APPROVED** [review]
> Reviewed 1 of 1 files at r9, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/koreader/koreader/9706)* status: all files reviewed, 1 unresolved discussion (waiting on @zwim)

**Review by NiLuJe — CHANGES_REQUESTED** [review]
I would get rid of the `.n` trickery, trying to outsmart LuaJIT is usually a fool's errand.

```diff
diff --git a/frontend/ui/uimanager.lua b/frontend/ui/uimanager.lua
index 917b67fd..13192121 100644
--- a/frontend/ui/uimanager.lua
+++ b/frontend/ui/uimanager.lua
@@ -31,7 +31,7 @@ local UIManager = {
 
     _now = time.now(),
     _window_stack = {},
-    _task_queue = { n = 0 },
+    _task_queue = {},
     _task_queue_dirty = false,
     _dirty = {},
     _zeromqs = {},
@@ -249,8 +249,7 @@ end
 
 -- Schedule an execution task; task queue is in descending order
 function UIManager:schedule(sched_time, action, ...)
-    local lo, hi = 1, self._task_queue.n
-    self._task_queue.n = hi + 1
+    local lo, hi = 1, #self._task_queue
     -- Leftmost binary insertion
     while lo <= hi do
         -- NOTE: We should be (mostly) free from overflow here, thanks to LuaJIT's BitOp semantics.
@@ -396,10 +395,9 @@ UIManager:unschedule(self.anonymousFunction)
 ]]
 function UIManager:unschedule(action)
     local removed = false
-    for i = self._task_queue.n, 1, -1 do
+    for i = #self._task_queue, 1, -1 do
         if self._task_queue[i].action == action then
             table.remove(self._task_queue, i)
-            self._task_queue.n = self._task_queue.n - 1
             removed = true
         end
     end
@@ -800,7 +798,7 @@ function UIManager:quit(exit_code)
     logger.info("quitting uimanager with exit code:", self._exit_code)
     self._task_queue_dirty = false
     self._window_stack = {}
-    self._task_queue = { n = 0 }
+    self._task_queue = {}
     for i = #self._zeromqs, 1, -1 do
         self._zeromqs[i]:stop()
     end
@@ -933,7 +931,7 @@ end
 
 --[[
 function UIManager:getNextTaskTimes(count)
-    count = math.min(count or 1, self._task_queue.n)
+    count = math.min(count or 1, #self._task_queue)
     local times = {}
     for i = 1, count do
         times[i] = self._task_queue[i].time - time.now()
@@ -943,7 +941,7 @@ end
 --]]
 
 function UIManager:getNextTaskTime()
-    local next_task = self._task_queue[self._task_queue.n]
+    local next_task = self._task_queue[#self._task_queue]
     if next_task then
         return next_task.time - time:now()
     end
@@ -956,12 +954,11 @@ function UIManager:_checkTasks()
     -- Tasks due for execution might themselves schedule more tasks (that might also be immediately due for execution ;)).
     -- Flipping this switch ensures we'll consume all such tasks *before* yielding to input polling.
     self._task_queue_dirty = false
-    while self._task_queue.n > 0 do
-        local task_time = self._task_queue[self._task_queue.n].time
+    while self._task_queue[1] do
+        local task_time = self._task_queue[#self._task_queue].time
         if task_time <= self._now then
             -- Remove the upcoming task, as it is due for execution...
             local task =  table.remove(self._task_queue)
-            self._task_queue.n = self._task_queue.n - 1
             -- ...so do it now.
             -- NOTE: Said task's action might modify _task_queue.
             --       To avoid race conditions and catch new upcoming tasks during this call,
diff --git a/spec/unit/taskqueue_bench.lua b/spec/unit/taskqueue_bench.lua
index 78857297..12b37853 100644
--- a/spec/unit/taskqueue_bench.lua
+++ b/spec/unit/taskqueue_bench.lua
@@ -11,7 +11,7 @@ describe("UIManager simple checkTasks and scheduling benchmark", function()
     local now = time.now()
     local wait_until -- luacheck: no unused
     UIManager:quit()
-    UIManager._task_queue = { n = 0 }
+    UIManager._task_queue = {}
 
     -- use schedule here, to be agnostic of the _task_queue order (ascending, descending).
     for i=1, NB_TESTS/2 do
@@ -46,7 +46,7 @@ describe("UIManager more advanced checkTasks and scheduling benchmark", function
     end
 
     for i=1, NB_TESTS do
-        UIManager._task_queue = { n = 0 }
+        UIManager._task_queue = {}
         UIManager:schedule(now + time.s(24*60*60), noop) -- shutdown
         UIManager:schedule(now + time.s(15*60*60), noop) -- sleep
         UIManager:schedule(now + time.s(55), noop) -- footer refresh
diff --git a/spec/unit/uimanager_bench.lua b/spec/unit/uimanager_bench.lua
index e23ea5cb..39f15f30 100644
--- a/spec/unit/uimanager_bench.lua
+++ b/spec/unit/uimanager_bench.lua
@@ -66,7 +66,7 @@ describe("UIManager more sophisticated schedule benchmark", function()
     end
 
     for i=1, NB_TESTS do
-        UIManager._task_queue = { n = 0 }
+        UIManager._task_queue = {}
         UIManager:schedule(now + time.s(24*60*60), noop) -- shutdown
         UIManager:schedule(now + time.s(15*60*60), noop) -- sleep
         UIManager:schedule(now + time.s(55), noop) -- footer refresh
@@ -91,7 +91,7 @@ describe("UIManager schedule massive collision tests", function()
 
     for i = 1, 6 do
         -- simple test (1000/10 collisions)
-        UIManager._task_queue = { n = 0 }
+        UIManager._task_queue = {}
         for j = 1, 10 do
             UIManager:schedule(math.random(10), j)
             -- check() -- enabling this takes really long O(n^2)
@@ -99,7 +99,7 @@ describe("UIManager schedule massive collision tests", function()
         check()
 
         -- armageddon test (10000 collisions)
-        UIManager._task_queue = { n = 0 }
+        UIManager._task_queue = {}
         for j = 1, 1e5 do
             UIManager:schedule(math.random(100), j)
             -- check() -- enabling this takes really long O(n^2)
@@ -114,7 +114,7 @@ describe("UIManager schedule massive ridiculous tests", function()
 
     for i = 1, 6 do
         -- simple test (1000 collisions)
-        UIManager._task_queue = { n = 0 }
+        UIManager._task_queue = {}
         local offs = 0
         for j = 1, 1e3 do
             UIManager:schedule(math.random(10), j + offs)
diff --git a/spec/unit/uimanager_spec.lua b/spec/unit/uimanager_spec.lua
index a98fbfaf..a01e67c9 100644
--- a/spec/unit/uimanager_spec.lua
+++ b/spec/unit/uimanager_spec.lua
@@ -21,7 +21,6 @@ describe("UIManager spec", function()
             { time = now - time.us(5), action = noop, args = {} },
             { time = now - time.s(10), action = noop, args = {} },
         }
-        UIManager._task_queue.n = #UIManager._task_queue
 
         UIManager:_checkTasks()
         assert.are.same(2, #UIManager._task_queue)
@@ -39,7 +38,6 @@ describe("UIManager spec", function()
             { time = now - time.us(5), action = noop, args = {} },
             { time = now - time.s(10), action = noop, args = {} },
         }
-        UIManager._task_queue.n = #UIManager._task_queue
 
         wait_until, now = UIManager:_checkTasks()
         assert.are.same(future_time, wait_until)
@@ -53,7 +51,6 @@ describe("UIManager spec", function()
             { time = now - time.us(5), action = noop, args = {} },
             { time = now - time.s(10), action = noop, args = {} },
         }
-        UIManager._task_queue.n = #UIManager._task_queue
 
         wait_until, now = UIManager:_checkTasks()
         assert.are.same(nil, wait_until)
@@ -75,7 +72,6 @@ describe("UIManager spec", function()
         UIManager._task_queue = {
             { time = future_time, action = '1', args = {} },
         }
-        UIManager._task_queue.n = #UIManager._task_queue
 
         assert.are.same(1, #UIManager._task_queue)
         UIManager:scheduleIn(150, 'quz')
@@ -86,7 +82,6 @@ describe("UIManager spec", function()
         UIManager._task_queue = {
             { time = now, action = '1', args = {} },
         }
-        UIManager._task_queue.n = #UIManager._task_queue
 
         assert.are.same(1, #UIManager._task_queue)
         UIManager:scheduleIn(150, 'foo')
@@ -105,7 +100,6 @@ describe("UIManager spec", function()
             { time = now - time.us(5), action = '2', args = {} },
             { time = now - time.s(10), action = '1', args = {} },
         }
-        UIManager._task_queue.n = #UIManager._task_queue
 
         -- insert into the tail slot
         UIManager:scheduleIn(10, 'foo')
@@ -186,7 +180,6 @@ describe("UIManager spec", function()
             { time = now - time.s(10), action = '1', args = {} },
             { time = now - time.s(15), action = '3', args = {} },
         }
-        UIManager._task_queue.n = #UIManager._task_queue
 
         -- insert into the tail slot
         UIManager:unschedule('3')
@@ -217,7 +210,6 @@ describe("UIManager spec", function()
             },
             { time = now - time.s(15), action = task_to_remove, args = {} }, -- this will be called
         }
-        UIManager._task_queue.n = #UIManager._task_queue
 
         UIManager:_checkTasks()
         assert.are.same(2, run_count)
```

This benches ever-so-slightly faster, and keeps the diff smaller, so, everybody wins ;).

**Review by NiLuJe — APPROVED** [review]
> Reviewed 4 of 4 files at r10, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/koreader/koreader/9706)* status: all files reviewed, 1 unresolved discussion (waiting on @zwim)
